### PR TITLE
Absent class miou fix

### LINKED ISSUE #2866: Zero value MeanIoU for Absent Classes in Ground Truth
## 🐛 Bug
In the context of multiclass semantic segmentation, the [torchmetrics.segmentation.MeanIoU](https://github.com/Lightning-AI/torchmetrics/blob/master/src/torchmetrics/segmentation/mean_iou.py#L29-L167) implementation with the `per_class=True` parameter, assigns an **Intersection over Union (IoU)** value of `0` for classes that are absent from the ground truth and correctly not predicted by the model. This behavior skews the **mean IoU** for those classes toward zero, even when predictions are perfect in cases where the class is present.

This can lead to misleading results for datasets with classes that are often absent from the images. The correct behavior should be to exclude such cases entirely when calculating the mean IoU for a class.

#### Code Example 1: Absent Class

```python
import torch
from torchmetrics.segmentation import MeanIoU

# Define the number of classes
num_classes = 3

# Initialize the MeanIoU metric
metric = MeanIoU(num_classes=num_classes, per_class=True, input_format='index')

# Example: Ground truth and predictions with class 2 absent
# preds and target are categorical masks (integer type)

target = torch.tensor([
    [0, 1],  # Ground truth: class 0, class 1
    [1, 0],  # Ground truth: class 1, class 0
])

preds = torch.tensor([
    [0, 1],  # Predictions: class 0, class 1
    [1, 0],  # Predictions: class 1, class 0
])

# Reset the metric (optional if not reused)
metric.reset()

# Update the metric with the predictions and target
metric.update(preds, target)

# Compute the mean IoU per class
miou_per_class = metric.compute()

# Print the results
for class_idx, iou in enumerate(miou_per_class):
    print(f"Class {class_idx} IoU: {iou:.4f}")
```

which output: 
```python
Class 0 IoU: 1.00
Class 1 IoU: 1.00
Class 2 IoU: 0.00
```
**Issue:** The IoU for Class 2 is 0.0, which is misleading because the class is absent and should not affect the IoU calculation.

#### Code Example 2: Perfect Predictions for All Classes
```python
import torch
from torchmetrics.segmentation import MeanIoU

# Define the number of classes
num_classes = 3

# Initialize the MeanIoU metric
metric = MeanIoU(num_classes=num_classes, per_class=True, input_format='index')
# Example: Ground truth and predictions perfectly consistent
# preds and target are categorical masks (integer type)
target = torch.tensor([
    [0, 1],  # Ground truth: class 0, class 1
    [1, 0],  # Ground truth: class 1, class 0
    [2, 2],  # Ground truth: class 2, class 2
])


preds = torch.tensor([
    [0, 1],  # Predictions: class 0, class 1
    [1, 0],  # Predictions: class 1, class 0
    [2, 2],  # Predictions: class 2, class 2
])

# Reset the metric (optional if not reused)
metric.reset()

# Update the metric with the predictions and target
metric.update(preds, target)

# Compute the mean IoU per class
miou_per_class = metric.compute()

# Print the results
for class_idx, iou in enumerate(miou_per_class):
    print(f"Class {class_idx} IoU: {iou:.4f}")
```
which output: 
```python
Class 0 IoU: 0.66
Class 1 IoU: 0.66
Class 2 IoU: 0.33
```
**Issue:** All classes are perfectly predicted, yet the IoUs are significantly below 1.0. This behavior is due to the fact that the IoU is zero when the class is absent. In fact in this case we obtain 0.66 ( 2/3) for classes 0 and 1 which are present in the first two samples but not in the third, and 0.33 (1/3) for class 2 present only in the last sample. 
Detailed computations for class 0, with 3 samples: 
```python
Class 0 Iou sample 0:  1 #perfectly predicted
Class 0 Iou sample 1:  1 #perfectly predicted
Class 0 Iou sample 2:  0 #since it is absent
Class 0 Mean IoU = (1+1+0) / 3 = 0.66
```
</details>

### Expected behavior

The expected behaviour should be to exclude samples that do not contain a specific class and for which that class was correctly not predicted, from the calculation of the mean Iou for that class.
A possible solution to this problem would be to implement ‘micro’ aggregation, where for each sample, the confusion matrix is simply updated and the IoU is calculated at the end, from the global confusion matrix.

### Environment

- TorchMetrics version: 1.5.2
- PyTorch version: 2.4.1+cu121
- Pytohn version: 3.10.15

### Additional context

Assigning 1 instead of 0 would not be a proper solution because it would bring the mean to be very close to 1, in a nutshell it would be a bit like reversing the problem. Especially if the dataset is very unbalanced and there are many images that do not contain a certain class, then the Mean IoU of that class will have a value very close to 0 or 1, depending on the value we decide to assign to the images that do not contain it. This prevents us from understanding the true performance of the model on the images in which that class is present, which are also the ones in which we are usually most interested.

### PR DESCRIPTION
**Pull request by Isalia20**
## What does this PR do?

Fixes #2866 

This is not the final version. I wanted to discuss the expected outputs for different edge cases and how to deal with them. I haven't included tests for these edge cases because monAI deals with such cases differently than what is proposed in this PR/issue, therefore adding a test and checking it against the monAI value would lead to failing tests. I'll post different inputs and what kind of outputs we should expect in the followup comment.

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**? **See above**

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun? partly 😄 

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Isalia20** [comment]
# Case 1: Perfect prediction
```
import torch
from torchmetrics.segmentation import MeanIoU
from torchmetrics.functional.segmentation.mean_iou import mean_iou

num_classes = 3
metric_one = MeanIoU(num_classes=num_classes, per_class=False, input_format='index')
metric_two = MeanIoU(num_classes=num_classes, per_class=True, input_format='index')
target = torch.tensor([
    [0, 1],  # Ground truth: class 0, class 1
    [1, 0],  # Ground truth: class 1, class 0
    [2, 2],  # Ground truth: class 2, class 2
])
preds = torch.tensor([
    [0, 1],  # Predictions: class 0, class 1
    [1, 0],  # Predictions: class 1, class 0
    [2, 2],  # Predictions: class 2, class 2
])
metric_one.update(preds, target)
miou_per_class_one = metric_one.compute()
metric_two.update(preds, target)
miou_per_class_two = metric_two.compute()

print(miou_per_class_one)
print(miou_per_class_two)

```
### Returns: 
```
tensor(1.)
tensor([1., 1., 1.])
```

# Case 2 Perfect prediction but one completely absent class
```
import torch
from torchmetrics.segmentation import MeanIoU
from torchmetrics.functional.segmentation.mean_iou import mean_iou

num_classes = 4
metric_one = MeanIoU(num_classes=num_classes, per_class=False, input_format='index')
metric_two = MeanIoU(num_classes=num_classes, per_class=True, input_format='index')
target = torch.tensor([
    [0, 1],  # Ground truth: class 0, class 1
    [1, 0],  # Ground truth: class 1, class 0
    [2, 2],  # Ground truth: class 2, class 2
])
preds = torch.tensor([
    [0, 1],  # Predictions: class 0, class 1
    [1, 0],  # Predictions: class 1, class 0
    [2, 2],  # Predictions: class 2, class 2
])
metric_one.update(preds, target)
miou_per_class_one = metric_one.compute()
metric_two.update(preds, target)
miou_per_class_two = metric_two.compute()

print(miou_per_class_one)
print(miou_per_class_two)
```

### Returns: 
```
tensor(nan)
tensor([1., 1., 1., nan])
```

Will update this with more cases soon

**Comment by Isalia20** [comment]
While it's marked WIP I still need comments from you @Borda . I posted some cases and wonder how I should proceed

**Comment by Isalia20** [comment]
Need a review on this @Borda

**Comment by Borda** [comment]
> Need a review on this @Borda

sure, let me have look next week :)

**Comment by SkafteNicki** [comment]
@Isalia20 no strong opinions from my side, but here is my take:
1. Agree that if there is a perfect match in input/output then the scores should be 1 
2. Agree that the correct value for the absent class in principal should be `nan` because we end up computing `0/0` in this case. However, we received feedback time and time again (for other metrics) that they want these cases to be ignored because one `nan` may kill a training process. So I am thinking that maybe the output should still be `tensor(1)` and instead a warning should be given to the users that nan values were computed but have been filtered
3. Agree that the value should be zero in this case

**Comment by Borda** [comment]
@Isalia20 @SkafteNicki do you think we could finish it for the next release this week? 🦩

**Comment by Isalia20** [comment]
One point here is that MonAI(which is used for validating the outputs of this) is used in tests. For such cases the behaviour is different from what I have listed in above cases. So my question is, can I use direct value comparison for such cases and disregard MonAI's output?

**Comment by Borda** [comment]
> One point here is that MonAI(which is used for validating the outputs of this) is used in tests. For such cases the behaviour is different from what I have listed in above cases. So my question is, can I use direct value comparison for such cases and disregard MonAI's output?

in general yes, we use validation with standard references or manually computed values...
do you have a specific one in mind?

**Comment by Isalia20** [comment]
Yes, I've put the cases in the above comment.

Here:

> # Case 1: Perfect prediction
> 
> ```python
> import torch
> from torchmetrics.segmentation import MeanIoU
> from torchmetrics.functional.segmentation.mean_iou import mean_iou
> 
> num_classes = 3
> metric_one = MeanIoU(num_classes=num_classes, per_class=False, input_format='index')
> metric_two = MeanIoU(num_classes=num_classes, per_class=True, input_format='index')
> target = torch.tensor([
>     [0, 1],  # Ground truth: class 0, class 1
>     [1, 0],  # Ground truth: class 1, class 0
>     [2, 2],  # Ground truth: class 2, class 2
> ])
> preds = torch.tensor([
>     [0, 1],  # Predictions: class 0, class 1
>     [1, 0],  # Predictions: class 1, class 0
>     [2, 2],  # Predictions: class 2, class 2
> ])
> metric_one.update(preds, target)
> miou_per_class_one = metric_one.compute()
> metric_two.update(preds, target)
> miou_per_class_two = metric_two.compute()
> 
> print(miou_per_class_one)
> print(miou_per_class_two)
> ```
> 
> ### Returns:
> 
> ```python
> tensor(1.)
> tensor([1., 1., 1.])
> ```
> 
> # Case 2 Perfect prediction but one completely absent class
> 
> ```python
> import torch
> from torchmetrics.segmentation import MeanIoU
> from torchmetrics.functional.segmentation.mean_iou import mean_iou
> 
> num_classes = 4
> metric_one = MeanIoU(num_classes=num_classes, per_class=False, input_format='index')
> metric_two = MeanIoU(num_classes=num_classes, per_class=True, input_format='index')
> target = torch.tensor([
>     [0, 1],  # Ground truth: class 0, class 1
>     [1, 0],  # Ground truth: class 1, class 0
>     [2, 2],  # Ground truth: class 2, class 2
> ])
> preds = torch.tensor([
>     [0, 1],  # Predictions: class 0, class 1
>     [1, 0],  # Predictions: class 1, class 0
>     [2, 2],  # Predictions: class 2, class 2
> ])
> metric_one.update(preds, target)
> miou_per_class_one = metric_one.compute()
> metric_two.update(preds, target)
> miou_per_class_two = metric_two.compute()
> 
> print(miou_per_class_one)
> print(miou_per_class_two)
> ```
> 
> ### Returns:
> 
> ```python
> tensor(nan)
> tensor([1., 1., 1., nan])
> ```
> 
> # Case 3: Completely wrong predictions(Probably same as the old one)
> 
> ```python
> import torch
> from torchmetrics.segmentation import MeanIoU
> 
> num_classes = 3
> metric_one = MeanIoU(num_classes=num_classes, per_class=False, input_format='index')
> metric_two = MeanIoU(num_classes=num_classes, per_class=True, input_format='index')
> target = torch.tensor([
>     [0, 1],  # Ground truth: class 0, class 1
>     [1, 0],  # Ground truth: class 1, class 0
>     [2, 2],  # Ground truth: class 2, class 2
> ])
> preds = torch.tensor([
>     [1, 2],  # Predictions: all wrong
>     [2, 1],  # Predictions: all wrong
>     [0, 1],  # Predictions: all wrong
> ])
> metric_one.update(preds, target)
> miou_per_class_one = metric_one.compute()
> metric_two.update(preds, target)
> miou_per_class_two = metric_two.compute()
> 
> print(miou_per_class_one)
> print(miou_per_class_two)
> ```
> 
> ### Returns
> 
> ```python
> tensor(0.)
> tensor([0., 0., 0.])
> ```

**Comment by Isalia20** [comment]
Ready for review, I think failing tests are unrelated. Instead of choosing to go with Nan's, I chose to go with -1.0's. If the user is requesting a single number(not per_class) then the class which wasn't present is ignored and iou is calculated for all other classes

**Comment by Isalia20** [comment]
And the tests are failing due to PyTDC here as well :/

File: src/torchmetrics/functional/segmentation/mean_iou.py [hunk 1]
```diff
@@ -127,15 +147,23 @@ def mean_iou(
...
     """
     _mean_iou_validate_args(num_classes, include_background, per_class, input_format)
+    preds, target = _mean_iou_reshape_args(preds, target, input_format)
     intersection, union = _mean_iou_update(preds, target, num_classes, include_background, input_format)
-    return _mean_iou_compute(intersection, union, per_class=per_class)
+    scores = _mean_iou_compute(intersection, union, zero_division="nan")
+    valid_classes = union > 0
+    if per_class:
+        return scores.nan_to_num(-1.0)
+    return scores.nansum(dim=-1) / valid_classes.sum(dim=-1)
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/segmentation/mean_iou.py:169]:
```suggestion
    return scores.nan_to_num(-1.0) if per_class else scores.nansum(dim=-1) / valid_classes.sum(dim=-1)
```

File: src/torchmetrics/functional/segmentation/mean_iou.py [hunk 2]
```diff
@@ -127,15 +147,23 @@ def mean_iou(
...
-                [0.2986, 0.3393, 0.0000, 0.0000, 0.0000],
-                [0.3515, 0.3432, 0.0000, 0.0000, 0.0000]])
+        tensor([[ 0.3617,  0.3617, -1.0000, -1.0000, -1.0000],
+                [ 0.3128,  0.3128, -1.0000, -1.0000, -1.0000],
+                [ 0.2727,  0.3366, -1.0000, -1.0000, -1.0000],
+                [ 0.3756,  0.3242, -1.0000, -1.0000, -1.0000]])
 
     """
     _mean_iou_validate_args(num_classes, include_background, per_class, input_format)
+    preds, target = _mean_iou_reshape_args(preds, target, input_format)
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/segmentation/mean_iou.py:163]: lets move this function call inside `_mean_iou_update`, then it should automatically also be called in the modular implementation which I guess it is not right now?

File: src/torchmetrics/segmentation/mean_iou.py [hunk 3]
```diff
@@ -90,7 +90,7 @@ class MeanIoU(Metric):
         tensor([0.3340, 0.3308])
         >>> miou = MeanIoU(num_classes=3, per_class=True, include_background=True, input_format="index")
         >>> miou(preds, target)
-        tensor([0.3334, 0.3336, 0.0000])
+        tensor([ 0.3334,  0.3336, -1.0000])
```
Comment (SkafteNicki) [inline src/torchmetrics/segmentation/mean_iou.py:93]: we should probably add in the docstring that a score of `-1` means absent class

File: tests/unittests/segmentation/test_mean_iou.py [hunk 4]
```diff
@@ -123,3 +123,56 @@ def test_mean_iou_functional(self, preds, target, input_format, num_classes, inc
                 "input_format": input_format,
             },
         )
+
+
+def test_mean_iou_absent_class():
+    """Test mean iou."""
```
Comment (SkafteNicki) [inline tests/unittests/segmentation/test_mean_iou.py:129]:
```suggestion
    """Test mean iou returns -1 for absent classes."""
```

File: src/torchmetrics/functional/segmentation/mean_iou.py [hunk 5]
```diff
@@ -127,15 +147,23 @@ def mean_iou(
...
+        >>> target = randint(0, 2, (4, 16, 16), generator=torch.Generator().manual_seed(43))
         >>> mean_iou(preds, target, num_classes=5, input_format = "index")
-        tensor([0.1331, 0.1369, 0.1276, 0.1389])
+        tensor([0.3617, 0.3128, 0.3047, 0.3499])
         >>> mean_iou(preds, target, num_classes=5, per_class=True, input_format="index")
-        tensor([[0.3330, 0.3323, 0.0000, 0.0000, 0.0000],
-                [0.3509, 0.3337, 0.0000, 0.0000, 0.0000],
-                [0.2986, 0.3393, 0.0000, 0.0000, 0.0000],
-                [0.3515, 0.3432, 0.0000, 0.0000, 0.0000]])
+        tensor([[ 0.3617,  0.3617, -1.0000, -1.0000, -1.0000],
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/segmentation/mean_iou.py:156]: we should probably add in the docstring that a score of -1 means absent class

File: src/torchmetrics/functional/segmentation/mean_iou.py (hunk 1 above)
Comment (Isalia20) [reply src/torchmetrics/functional/segmentation/mean_iou.py]: fixed

File: src/torchmetrics/functional/segmentation/mean_iou.py (hunk 2 above)
Comment (Isalia20) [reply src/torchmetrics/functional/segmentation/mean_iou.py]: fixed

File: src/torchmetrics/segmentation/mean_iou.py (hunk 3 above)
Comment (Isalia20) [reply src/torchmetrics/segmentation/mean_iou.py:94]: added

File: tests/unittests/segmentation/test_mean_iou.py (hunk 4 above)
Comment (Isalia20) [reply tests/unittests/segmentation/test_mean_iou.py]: added

File: src/torchmetrics/functional/segmentation/mean_iou.py (hunk 5 above)
Comment (Isalia20) [reply src/torchmetrics/functional/segmentation/mean_iou.py:159]: added

**Comment by Isalia20** [comment]
resolved the comments, should be good to merge. Those 2 tests failed on purging cached refs stage

**Review by SkafteNicki — APPROVED** [review]

**Review by Borda — APPROVED** [review]
